### ISSUE TITLE
Add React 18 strict mode support (WAS: React 18 nested state "set" method crash in development mode.)

### ISSUE THREAD

**Issue by Chill-Studio** [avkonst/hookstate#305]
Hello,

I use react 18.
The example I pasted the documentation from the doc [https://hookstate.js.org/docs/nested-state] causes a crash in **development mode ONLY**, it work well when you build it though.

It looks like a state is destroyed (maybe because of the double useEffect that happens only in dev mode) which causes the "set" call to throw an error.

This is what you can see in the devtool on startup before any action ( notice the destroy )

![error1](https://user-images.githubusercontent.com/66690474/177390540-3b24cd7b-855b-46c0-80d7-d2990ae59940.png)

Here is the error when clicking on "Add a task"

![2](https://user-images.githubusercontent.com/66690474/177390537-ca9fc52d-7bca-47b3-a6fd-273b709c0576.png)

**Comment by avkonst** [comment]
Hi, what is the version of hookstate? Could you please post the full app component starting from the index.tsx file?

**Comment by Chill-Studio** [comment]
"@hookstate/core": "^4.0.0-rc11",
    "@hookstate/devtools": "^3.0.2",
```
    
    
    The strict minimum
    
    
    import React from 'react';
import ReactDOM from 'react-dom/client';
import App from './App';

import './index.css';

ReactDOM.createRoot(document.getElementById('root')!).render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
);

```

and the guilty part
```

import React from 'react';
import { useState, State } from '@hookstate/core';

interface Task {
  name: string;
  priority?: number;
}

export default function App() {
  const state: State<Task[]> = useState([{ name: 'First Task' }] as Task[]);
  return (
    <>
      {state.map((taskState: State<Task>, taskIndex) => (
        <TaskEditor key={taskIndex} taskState={taskState} />
      ))}
      <button onClick={() => state.merge([{ name: 'Untitled' }])}>Add task</button>
    </>
  );
}

function TaskEditor(props: { taskState: State<Task> }) {
  const taskState = props.taskState;
  return (
    <p>
      <input value={taskState.name.get()} onChange={(e) => taskState.name.set(e.target.value)} />
    </p>
  );
}

```

**Comment by avkonst** [comment]
This is because React 18 introduced this thing to the strict mode. https://reactjs.org/docs/strict-mode.html#ensuring-reusable-state They unmount the component but do not destroy the state. Hookstate assumes that unmounting a component means the same state is never reused again and is actually destroyed. Some plugins rely on destroy to release the resources, like close websocket, remove subscriptions, etc... This is bizar what Facebook is doing...

**Comment by avkonst** [comment]
Fixed in @hookstate/core@4.0.0-rc12, please verify.
